File: backend/F1Fantasy/Controllers/PredictionsController.cs
using F1Fantasy.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;

namespace F1Fantasy.Controllers;

[Authorize]
[ApiController]
[Route("api/[controller]")]
public class PredictionsController : ControllerBase
{
    private readonly PredictionService _predictionService;

    public PredictionsController(PredictionService predictionService)
    {
        _predictionService = predictionService;
    }

    private string GetUserId()
    {
        return User.FindFirstValue(ClaimTypes.NameIdentifier) 
            ?? throw new UnauthorizedAccessException("User ID not found");
    }

    // Constructor Championship
    [HttpPost("groups/{groupId}/constructor-championship")]
    public async Task<IActionResult> SaveConstructorChampionship(int groupId, [FromBody] List<string> rankedConstructorIds)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.SaveConstructorChampionshipAsync(groupId, userId, rankedConstructorIds);
            return Ok(prediction);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/constructor-championship")]
    public async Task<IActionResult> GetConstructorChampionship(int groupId)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.GetConstructorChampionshipAsync(groupId, userId);
            if (prediction == null) return NotFound();
            return Ok(prediction);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    // Driver Championship
    [HttpPost("groups/{groupId}/driver-championship")]
    public async Task<IActionResult> SaveDriverChampionship(int groupId, [FromBody] List<string> rankedDriverIds)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.SaveDriverChampionshipAsync(groupId, userId, rankedDriverIds);
            return Ok(prediction);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/driver-championship")]
    public async Task<IActionResult> GetDriverChampionship(int groupId)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.GetDriverChampionshipAsync(groupId, userId);
            if (prediction == null) return NotFound();
            return Ok(prediction);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    // Driver Draft
    [HttpPost("groups/{groupId}/driver-draft")]
    public async Task<IActionResult> SaveDriverDraft(int groupId, [FromBody] DriverDraftRequest request)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.SaveDriverDraftAsync(groupId, userId, request.Driver1Id, request.Driver2Id);
            return Ok(prediction);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/driver-draft")]
    public async Task<IActionResult> GetDriverDraft(int groupId)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.GetDriverDraftAsync(groupId, userId);
            if (prediction == null) return NotFound();
            return Ok(prediction);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    // Destructor
    [HttpPost("groups/{groupId}/destructor")]
    public async Task<IActionResult> SaveDestructor(int groupId, [FromBody] TwoDriverRequest request)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.SaveDestructorAsync(groupId, userId, request.Driver1Id, request.Driver2Id);
            return Ok(prediction);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/destructor")]
    public async Task<IActionResult> GetDestructor(int groupId)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.GetDestructorAsync(groupId, userId);
            if (prediction == null) return NotFound();
            return Ok(prediction);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    // Mr Saturday
    [HttpPost("groups/{groupId}/mr-saturday")]
    public async Task<IActionResult> SaveMrSaturday(int groupId, [FromBody] TwoDriverRequest request)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.SaveMrSaturdayAsync(groupId, userId, request.Driver1Id, request.Driver2Id);
            return Ok(prediction);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/mr-saturday")]
    public async Task<IActionResult> GetMrSaturday(int groupId)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.GetMrSaturdayAsync(groupId, userId);
            if (prediction == null) return NotFound();
            return Ok(prediction);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    // Zero Pointer
    [HttpPost("groups/{groupId}/zero-pointer")]
    public async Task<IActionResult> SaveZeroPointer(int groupId, [FromBody] TwoDriverRequest request)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.SaveZeroPointerAsync(groupId, userId, request.Driver1Id, request.Driver2Id);
            return Ok(prediction);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/zero-pointer")]
    public async Task<IActionResult> GetZeroPointer(int groupId)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.GetZeroPointerAsync(groupId, userId);
            if (prediction == null) return NotFound();
            return Ok(prediction);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    // Wildcard
    [HttpPost("groups/{groupId}/wildcard")]
    public async Task<IActionResult> SaveWildcard(int groupId, [FromBody] WildcardRequest request)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.SaveWildcardAsync(groupId, userId, request.Statement);
            return Ok(prediction);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(new { error = ex.Message });
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }

    [HttpGet("groups/{groupId}/wildcard")]
    public async Task<IActionResult> GetWildcard(int groupId)
    {
        try
        {
            var userId = GetUserId();
            var prediction = await _predictionService.GetWildcardAsync(groupId, userId);
            if (prediction == null) return NotFound();
            return Ok(prediction);
        }
        catch (Exception ex)
        {
            return BadRequest(new { error = ex.Message });
        }
    }
}

public record DriverDraftRequest(string? Driver1Id, string? Driver2Id);
public record TwoDriverRequest(string? Driver1Id, string? Driver2Id);
public record WildcardRequest(string? Statement);

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/controller/PredictionsController.java
package no.f1fantasy.controller;

import java.security.Principal;
import java.util.List;

import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import no.f1fantasy.dto.DriverDraftRequest;
import no.f1fantasy.dto.ErrorResponse;
import no.f1fantasy.dto.TwoDriverRequest;
import no.f1fantasy.entity.ConstructorChampionshipPrediction;
import no.f1fantasy.entity.DestructorPrediction;
import no.f1fantasy.entity.DriverChampionshipPrediction;
import no.f1fantasy.entity.DriverDraftPrediction;
import no.f1fantasy.entity.MrSaturdayPrediction;
import no.f1fantasy.entity.WildcardPrediction;
import no.f1fantasy.entity.ZeroPointerPrediction;
import no.f1fantasy.service.PredictionService;

@RestController
@RequestMapping("/api/predictions")
@PreAuthorize("isAuthenticated()")
@SuppressWarnings("null")
public class PredictionsController {
    
    private final PredictionService predictionService;

    public PredictionsController(PredictionService predictionService) {
        this.predictionService = predictionService;
    }

    private String getUserId(Principal principal) {
        if (principal == null) {
            throw new IllegalArgumentException("User ID not found");
        }
        return principal.getName();
    }

    // Constructor Championship
    @PostMapping("/groups/{groupId}/constructor-championship")
    public ResponseEntity<?> saveConstructorChampionship(
            @PathVariable int groupId,
            @RequestBody List<String> rankedConstructorIds,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            ConstructorChampionshipPrediction prediction = predictionService.saveConstructorChampionship(groupId, userId, rankedConstructorIds);
            return ResponseEntity.ok(prediction);
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/constructor-championship")
    public ResponseEntity<?> getConstructorChampionship(
            @PathVariable int groupId,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            var prediction = predictionService.getConstructorChampionship(groupId, userId);
            if (prediction.isEmpty()) {
                return ResponseEntity.notFound().build();
            }
            return ResponseEntity.ok(prediction.get());
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    // Driver Championship
    @PostMapping("/groups/{groupId}/driver-championship")
    public ResponseEntity<?> saveDriverChampionship(
            @PathVariable int groupId,
            @RequestBody List<String> rankedDriverIds,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            DriverChampionshipPrediction prediction = predictionService.saveDriverChampionship(groupId, userId, rankedDriverIds);
            return ResponseEntity.ok(prediction);
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/driver-championship")
    public ResponseEntity<?> getDriverChampionship(
            @PathVariable int groupId,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            var prediction = predictionService.getDriverChampionship(groupId, userId);
            if (prediction.isEmpty()) {
                return ResponseEntity.notFound().build();
            }
            return ResponseEntity.ok(prediction.get());
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    // Driver Draft
    @PostMapping("/groups/{groupId}/driver-draft")
    public ResponseEntity<?> saveDriverDraft(
            @PathVariable int groupId,
            @RequestBody DriverDraftRequest request,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            DriverDraftPrediction prediction = predictionService.saveDriverDraft(groupId, userId, request.getDriver1Id(), request.getDriver2Id());
            return ResponseEntity.ok(prediction);
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/driver-draft")
    public ResponseEntity<?> getDriverDraft(
            @PathVariable int groupId,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            var prediction = predictionService.getDriverDraft(groupId, userId);
            if (prediction.isEmpty()) {
                return ResponseEntity.notFound().build();
            }
            return ResponseEntity.ok(prediction.get());
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    // Destructor
    @PostMapping("/groups/{groupId}/destructor")
    public ResponseEntity<?> saveDestructor(
            @PathVariable int groupId,
            @RequestBody TwoDriverRequest request,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            DestructorPrediction prediction = predictionService.saveDestructor(groupId, userId, request.getDriver1Id(), request.getDriver2Id());
            return ResponseEntity.ok(prediction);
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/destructor")
    public ResponseEntity<?> getDestructor(
            @PathVariable int groupId,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            var prediction = predictionService.getDestructor(groupId, userId);
            if (prediction.isEmpty()) {
                return ResponseEntity.notFound().build();
            }
            return ResponseEntity.ok(prediction.get());
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    // Mr Saturday
    @PostMapping("/groups/{groupId}/mr-saturday")
    public ResponseEntity<?> saveMrSaturday(
            @PathVariable int groupId,
            @RequestBody TwoDriverRequest request,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            MrSaturdayPrediction prediction = predictionService.saveMrSaturday(groupId, userId, request.getDriver1Id(), request.getDriver2Id());
            return ResponseEntity.ok(prediction);
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/mr-saturday")
    public ResponseEntity<?> getMrSaturday(
            @PathVariable int groupId,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            var prediction = predictionService.getMrSaturday(groupId, userId);
            if (prediction.isEmpty()) {
                return ResponseEntity.notFound().build();
            }
            return ResponseEntity.ok(prediction.get());
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    // Zero Pointer
    @PostMapping("/groups/{groupId}/zero-pointer")
    public ResponseEntity<?> saveZeroPointer(
            @PathVariable int groupId,
            @RequestBody List<String> driverIds,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            ZeroPointerPrediction prediction = predictionService.saveZeroPointer(groupId, userId, driverIds);
            return ResponseEntity.ok(prediction);
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/zero-pointer")
    public ResponseEntity<?> getZeroPointer(
            @PathVariable int groupId,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            var prediction = predictionService.getZeroPointer(groupId, userId);
            if (prediction.isEmpty()) {
                return ResponseEntity.notFound().build();
            }
            return ResponseEntity.ok(prediction.get());
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    // Wildcard
    @PostMapping("/groups/{groupId}/wildcard")
    public ResponseEntity<?> saveWildcard(
            @PathVariable int groupId,
            @RequestBody String statement,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            WildcardPrediction prediction = predictionService.saveWildcard(groupId, userId, statement);
            return ResponseEntity.ok(prediction);
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/wildcard")
    public ResponseEntity<?> getWildcard(
            @PathVariable int groupId,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            var prediction = predictionService.getWildcard(groupId, userId);
            if (prediction.isEmpty()) {
                return ResponseEntity.notFound().build();
            }
            return ResponseEntity.ok(prediction.get());
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }

    @GetMapping("/groups/{groupId}/wildcards")
    public ResponseEntity<?> getAllWildcards(
            @PathVariable int groupId,
            Principal principal) {
        try {
            String userId = getUserId(principal);
            List<WildcardPrediction> predictions = predictionService.getAllWildcards(groupId, userId);
            return ResponseEntity.ok(predictions);
        } catch (Exception ex) {
            return ResponseEntity.badRequest().body(new ErrorResponse(ex.getMessage()));
        }
    }
}
